tests/unit_tests/blas/level1/rotg_usm.cpp

```diff
@@ -86,15 +86,29 @@ int test(device *dev, oneapi::mkl::layout layout) {
     ::rotg((fp_ref *)&a_ref, (fp_ref *)&b_ref, (fp_scalar *)&c_ref, (fp_ref *)&s_ref);
 
     // Call DPC++ ROTG.
-    fp *a_p = (fp *)oneapi::mkl::malloc_shared(64, sizeof(fp), *dev, cxt);
-    fp *b_p = (fp *)oneapi::mkl::malloc_shared(64, sizeof(fp), *dev, cxt);
-    fp *s_p = (fp *)oneapi::mkl::malloc_shared(64, sizeof(fp), *dev, cxt);
-    fp_scalar *c_p = (fp_scalar *)oneapi::mkl::malloc_shared(64, sizeof(fp_scalar), *dev, cxt);
+    fp *a_p, *b_p, *s_p;
+    fp_scalar *c_p;
+    if constexpr (alloc_type == usm::alloc::shared) {
+        a_p = (fp *)oneapi::mkl::malloc_shared(64, sizeof(fp), *dev, cxt);
+        b_p = (fp *)oneapi::mkl::malloc_shared(64, sizeof(fp), *dev, cxt);
+        s_p = (fp *)oneapi::mkl::malloc_shared(64, sizeof(fp), *dev, cxt);
+        c_p = (fp_scalar *)oneapi::mkl::malloc_shared(64, sizeof(fp_scalar), *dev, cxt);
+    }
+    else if constexpr (alloc_type == usm::alloc::device) {
+        a_p = (fp *)oneapi::mkl::malloc_device(64, sizeof(fp), *dev, cxt);
+        b_p = (fp *)oneapi::mkl::malloc_device(64, sizeof(fp), *dev, cxt);
+        s_p = (fp *)oneapi::mkl::malloc_device(64, sizeof(fp), *dev, cxt);
+        c_p = (fp_scalar *)oneapi::mkl::malloc_device(64, sizeof(fp_scalar), *dev, cxt);
+    }
+    else {
+        throw std::runtime_error("Bad alloc_type");
+    }
 
-    a_p[0] = a;
-    b_p[0] = b;
-    s_p[0] = s;
-    c_p[0] = c;
+    main_queue.memcpy(a_p, &a, sizeof(fp)).wait();
+    main_queue.memcpy(b_p, &b, sizeof(fp)).wait();
+    main_queue.memcpy(s_p, &s, sizeof(fp)).wait();
+    main_queue.memcpy(c_p, &c, sizeof(fp_scalar)).wait();
```

Comment:
Can we avoid these if `malloc_shared` is used? Also no need for multiple `wait()`. One at the end is sufficient.

Reply:
Agreed on the `.wait()` (fixed). We can certainly avoid these if `malloc_shared` is used, but the point of the new tests is to test what happens with device memory, which caused a segfault before this PR.